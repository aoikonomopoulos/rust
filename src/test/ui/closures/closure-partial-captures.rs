// compile-pass
// ignore-test
#![feature(rustc_attrs)]

#[derive(Default)]
struct D {
    d: usize,
}

#[derive(Default)]
struct B {
    b1: D,
}

#[derive(Default)]
struct C {
    c1: D,
    c2: D,
}

#[derive(Default)]
struct A {
    a1: B,
    a2: C,
}

fn ref_imm(_arg: &usize) {}
fn ref_mut(_arg: &mut usize) {}

#[rustc_dump_closure_captures]
fn interior_simple() {
    let interior: A = Default::default();
    let _c = || {
        // -^ NOTE Upvar local interior CapturePath([Field(a1), Field(b1), Field(d)]): ByRef immutable
        interior.a1.b1.d + 1
    };
}

#[rustc_dump_closure_captures]
fn interior_full_immutable() {
    let interior: A = Default::default();
    let _c = || {
        // -^ NOTE Upvar local interior CapturePath([Field(a2)]): ByRef immutable
        interior.a2.c1.d + interior.a2.c2.d
    };
}

#[rustc_dump_closure_captures]
fn interior_full_mutable() {
    let mut interior1: A = Default::default();
    let mut interior2: A = Default::default();
    let _c1 = || {
        // -^ NOTE Upvar local interior1 CapturePath([Field(a2)]): ByRef mutable
        ref_imm(&interior1.a2.c1.d);
        ref_mut(&mut interior1.a2.c2.d);
    };
    let _c2 = || {
        // -^ NOTE Upvar local interior1 CapturePath([Field(a2)]): ByRef mutable
        ref_mut(&mut interior2.a2.c2.d);
        ref_imm(&interior2.a2.c1.d);
    };
}

#[rustc_dump_closure_captures]
fn capture_borrow() {
    let cb = 6;
    let _c = || {
        // -^ NOTE Upvar local cb CapturePath([]): ByRef immutable
        ref_imm(&cb);
    };
}

#[rustc_dump_closure_captures]
fn capture_borrow_mut() {
    let mut cbm = 6;
    let _c = || {
        // -^ NOTE Upvar local cbm CapturePath([]): ByRef mutable
        cbm += 1
    };
}

#[rustc_dump_closure_captures]
fn capture_borrow_unique() {
    let mut target = 7;
    let cbu = &mut target;
    let _c = || {
        // -^ NOTE Upvar local cbu CapturePath([]): ByRef unique immutable
        *cbu += 1
    };
}

#[rustc_dump_closure_captures]
fn capture_move() {
    let cm = 7;
    let _c = move || {
        // -^ NOTE Upvar local cm CapturePath([]): ByValue
        cm + 1
    };
}

#[rustc_dump_closure_captures]
fn capture_move_mut() {
    let mut cmm = 8;
    let _c = move || {
        // -^ NOTE Upvar local cb CapturePath([]): ByValue
        cmm += 1
    };
}

#[rustc_dump_closure_captures]
fn capture_pat() {
    let interior1: A = Default::default();
    let _c = || {
        // -^ NOTE Upvar local interior1 CapturePath([Field(a1)]): ByValue
        match interior1 {
            A {a1, ..} => { // This hits `upvar::consume_pat`
                ref_imm(&a1.b1.d);
            }
        }
    };
    let interior2: A = Default::default();
    let _c = || {
        // -^ NOTE Upvar local interior2 CapturePath([Field(a2)]): ByRef immutable
        match interior2 {
            A {ref a2, ..} => {
                ref_imm(&a2.c1.d);
            }
        }
    };
    let mut interior3: A = Default::default();
    let _c = || {
        // -^ NOTE Upvar local mut interior3 CapturePath([Field(a1)]): ByRef mutable
        match interior3 {
            A {ref mut a1, ..} => {
                ref_mut(&mut a1.b1.d);
            }
        }
    };
}

// The UpvarCapture can be
// - by value
// - by ref, for which the borrow kind may be
//   - Immutable
//   - Unique Immutable
//   - Mutable
fn main() {
    interior_simple();
    // interior_full_immutable();
    // interior_full_mutable();
    // capture_borrow();
    // capture_borrow_mut();
    // capture_borrow_unique();
    // capture_move();
    // capture_move_mut();
    // capture_pat();
}
