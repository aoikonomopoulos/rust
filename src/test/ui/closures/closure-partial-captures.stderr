note: closure capture path
  --> $DIR/closure-partial-captures.rs:32:14
   |
LL |       let _c = || {
   |  ______________^
LL | |         // -^ NOTE Upvar local interior CapturePath([Field(a1), Field(b1), Field(d)]): ByRef immutable
LL | |         interior.a1.b1.d + 1
LL | |     };
   | |_____^
   |
   = note: Upvar local interior CapturePath([Field(a1), Field(b1), Field(d)]): ByRef immutable

note: closure capture path
  --> $DIR/closure-partial-captures.rs:41:14
   |
LL |       let _c = || {
   |  ______________^
LL | |         // -^ NOTE Upvar local interior CapturePath([Field(a2)]): ByRef immutable
LL | |         interior.a2.c1.d + interior.a2.c2.d
LL | |     };
   | |_____^
   |
   = note: Upvar local interior CapturePath([Field(a2)]): ByRef immutable

note: closure capture path
  --> $DIR/closure-partial-captures.rs:51:15
   |
LL |       let _c1 = || {
   |  _______________^
LL | |         // -^ NOTE Upvar local interior1 CapturePath([Field(a2)]): ByRef mutable
LL | |         ref_imm(&interior1.a2.c1.d);
LL | |         ref_mut(&mut interior1.a2.c2.d);
LL | |     };
   | |_____^
   |
   = note: Upvar local mut interior1 CapturePath([Field(a2)]): ByRef mutable

note: closure capture path
  --> $DIR/closure-partial-captures.rs:56:15
   |
LL |       let _c2 = || {
   |  _______________^
LL | |         // -^ NOTE Upvar local interior1 CapturePath([Field(a2)]): ByRef mutable
LL | |         ref_mut(&mut interior2.a2.c2.d);
LL | |         ref_imm(&interior2.a2.c1.d);
LL | |     };
   | |_____^
   |
   = note: Upvar local mut interior2 CapturePath([Field(a2)]): ByRef mutable

note: closure capture path
  --> $DIR/closure-partial-captures.rs:66:14
   |
LL |       let _c = || {
   |  ______________^
LL | |         // -^ NOTE Upvar local cb CapturePath([]): ByRef immutable
LL | |         ref_imm(&cb);
LL | |     };
   | |_____^
   |
   = note: Upvar local cb CapturePath([]): ByRef immutable

note: closure capture path
  --> $DIR/closure-partial-captures.rs:75:14
   |
LL |       let _c = || {
   |  ______________^
LL | |         // -^ NOTE Upvar local cbm CapturePath([]): ByRef mutable
LL | |         cbm += 1
LL | |     };
   | |_____^
   |
   = note: Upvar local mut cbm CapturePath([]): ByRef mutable

note: closure capture path
  --> $DIR/closure-partial-captures.rs:85:14
   |
LL |       let _c = || {
   |  ______________^
LL | |         // -^ NOTE Upvar local cbu CapturePath([]): ByRef unique immutable
LL | |         *cbu += 1
LL | |     };
   | |_____^
   |
   = note: Upvar local cbu CapturePath([]): ByRef unique immutable

note: closure capture path
  --> $DIR/closure-partial-captures.rs:94:14
   |
LL |       let _c = move || {
   |  ______________^
LL | |         // -^ NOTE Upvar local cm CapturePath([]): ByValue
LL | |         cm + 1
LL | |     };
   | |_____^
   |
   = note: Upvar local cm CapturePath([]): ByValue

note: closure capture path
  --> $DIR/closure-partial-captures.rs:103:14
   |
LL |       let _c = move || {
   |  ______________^
LL | |         // -^ NOTE Upvar local cb CapturePath([]): ByValue
LL | |         cmm += 1
LL | |     };
   | |_____^
   |
   = note: Upvar local mut cmm CapturePath([]): ByValue

note: closure capture path
  --> $DIR/closure-partial-captures.rs:112:14
   |
LL |       let _c = || {
   |  ______________^
LL | |         // -^ NOTE Upvar local interior1 CapturePath([Field(a1)]): ByValue
LL | |         match interior1 {
LL | |             A {a1, ..} => { // This hits `upvar::consume_pat`
...  |
LL | |         }
LL | |     };
   | |_____^
   |
   = note: Upvar local interior1 CapturePath([Field(a1)]): ByValue

note: closure capture path
  --> $DIR/closure-partial-captures.rs:121:14
   |
LL |       let _c = || {
   |  ______________^
LL | |         // -^ NOTE Upvar local interior2 CapturePath([Field(a2)]): ByRef immutable
LL | |         match interior2 {
LL | |             A {ref a2, ..} => {
...  |
LL | |         }
LL | |     };
   | |_____^
   |
   = note: Upvar local interior2 CapturePath([Field(a2)]): ByRef immutable

note: closure capture path
  --> $DIR/closure-partial-captures.rs:130:14
   |
LL |       let _c = || {
   |  ______________^
LL | |         // -^ NOTE Upvar local mut interior3 CapturePath([Field(a1)]): ByRef mutable
LL | |         match interior3 {
LL | |             A {ref mut a1, ..} => {
...  |
LL | |         }
LL | |     };
   | |_____^
   |
   = note: Upvar local mut interior3 CapturePath([Field(a1)]): ByRef mutable


